### DISCUSSION TITLE
Lost in Errors (Beginner)

### DISCUSSION THREAD

**Discussion by NOCTish** [theOehrly/Fast-F1#327] [Q&A]
Greetings!
I've installed the latest version if FastF1 Library and requirements.
problem is no matter what code I try to run, there is huge amount of errors which I can not resolve. please help me!
```
import fastf1 as ff1
import os
script_dir = os.path.dirname(os.path.abspath(__file__))
cache_dir = os.path.join(script_dir, 'cache')
if not os.path.exists(cache_dir):
    os.makedirs(cache_dir)
ff1.Cache.enable_cache(cache_dir)

YEAR = 2022
RACE = 'Bahrain'
SESSION = 'R'

session = ff1.get_session(
    YEAR,
    RACE,
    SESSION
)

session.load()
laps = session.laps
laps.pick_fastest()
laps.pick_driver('HAM')

alo_laps = laps.pick_driver('ALO')
alo_fastest = alo_laps.pick_fastest()

alo_fastest.get_car_data()
```

**Here is the errors:**
PS C:\Users\amirm> & C:/Users/amirm/AppData/Local/Programs/Python/Python310/python.exe "c:/Users/amirm/Desktop/F1/import fastf1 as ff1.py"                                                               core           INFO     Loading data for Bahrain Grand Prix - Race [v2.3.0]
api            INFO     No cached data found for driver_info. Loading data...
api            INFO     Fetching driver list...
core        WARNING     Failed to load extended driver information!
api            INFO     No cached data found for timing_data. Loading data...
api            INFO     Fetching timing data...
core        WARNING     Failed to load lap data!
api            INFO     No cached data found for car_data. Loading data...
api            INFO     Fetching car data...
core        WARNING     Failed to load telemetry data!
api            INFO     No cached data found for weather_data. Loading data...
api            INFO     Fetching weather data...
core        WARNING     Failed to load weather data!
api            INFO     No cached data found for race_control_messages. Loading data...
api            INFO     Fetching race control messages...
core        WARNING     Failed to load Race Control message data!
core           INFO     Finished loading data for 20 drivers: ['16', '55', '44', '63', '20', '77', '31', '22', '14', '24', '47', '18', '23', '3', '4', '6', '27', '11', '1', '10']
Traceback (most recent call last):
  File "c:\Users\amirm\Desktop\F1\import fastf1 as ff1.py", line 20, in <module>
    laps = session.laps
  File "C:\Users\amirm\AppData\Local\Programs\Python\Python310\lib\site-packages\fastf1\core.py", line 1052, in laps
    return self._get_property_warn_not_loaded('_laps')
  File "C:\Users\amirm\AppData\Local\Programs\Python\Python310\lib\site-packages\fastf1\core.py", line 1017, in _get_property_warn_not_loaded
    raise DataNotLoadedError("The data you are trying to access has not "
fastf1.core.DataNotLoadedError: The data you are trying to access has not been loaded yet. See `Session.load`

**Comment by theOehrly** [comment]
Data should be available. Does that issue still occur? There seemed to be some issues with the API somewhere in the last few days. Maybe you were just unlucky.

**Comment by NOCTish** [comment]
Unfortunetly Yes. I still get the core & api errors. I ran this in Jupyter Notebook even
RN and here are the errors:
 **CODE:**

```
import fastf1 as ff1
ff1.Cache.enable_cache('cache')

quali_session = ff1.get_session(2022,'Bahrain','Q')
quali_session.load()
```

```
core           INFO 	Loading data for Bahrain Grand Prix - Qualifying [v2.3.0]
api            INFO 	No cached data found for driver_info. Loading data...
api            INFO 	Fetching driver list...
core        WARNING 	Failed to load extended driver information!
api            INFO 	No cached data found for timing_data. Loading data...
api            INFO 	Fetching timing data...
core        WARNING 	Failed to load lap data!
api            INFO 	No cached data found for car_data. Loading data...
api            INFO 	Fetching car data...
core        WARNING 	Failed to load telemetry data!
api            INFO 	No cached data found for weather_data. Loading data...
api            INFO 	Fetching weather data...
core        WARNING 	Failed to load weather data!
api            INFO 	No cached data found for race_control_messages. Loading data...
api            INFO 	Fetching race control messages...
core        WARNING 	Failed to load Race Control message data!
core           INFO 	Finished loading data for 20 drivers: ['16', '1', '55', '11', '44', '77', '20', '14', '63', '10', '31', '47', '4', '23', '24', '22', '27', '3', '18', '6']
```

**Comment by NOCTish** [comment]
further info
I ran the code with anaconda notebook
```
import fastf1 as ff1

ff1.Cache.enable_cache('cache')

quali_session = ff1.get_session(2022,'Bahrain','Q')
quali_session.load()
quali_session.laps
```

**Core Errors:**
```
core           INFO 	Loading data for Bahrain Grand Prix - Qualifying [v2.3.0]
api            INFO 	No cached data found for driver_info. Loading data...
api            INFO 	Fetching driver list...
core        WARNING 	Failed to load extended driver information!
api            INFO 	No cached data found for timing_data. Loading data...
api            INFO 	Fetching timing data...
core        WARNING 	Failed to load lap data!
api            INFO 	No cached data found for car_data. Loading data...
api            INFO 	Fetching car data...
core        WARNING 	Failed to load telemetry data!
api            INFO 	No cached data found for weather_data. Loading data...
api            INFO 	Fetching weather data...
core        WARNING 	Failed to load weather data!
api            INFO 	No cached data found for race_control_messages. Loading data...
api            INFO 	Fetching race control messages...
core        WARNING 	Failed to load Race Control message data!
core           INFO 	Finished loading data for 20 drivers: ['16', '1', '55', '11', '44', '77', '20', '14', '63', '10', '31', '47', '4', '23', '24', '22', '27', '3', '18', '6']
```

**and then, session laps errors:**

```
---------------------------------------------------------------------------
DataNotLoadedError                        Traceback (most recent call last)
Cell In [6], line 7
      5 quali_session = ff1.get_session(2022,'Bahrain','Q')
      6 quali_session.load()
----> 7 quali_session.laps

File ~\AppData\Local\Programs\Python\Python310\lib\site-packages\fastf1\core.py:1052, in Session.laps(self)
   1046 @property
   1047 def laps(self):
   1048     """:class:`Laps`: All laps from all drivers driven in this session.
   1049 
   1050     Data is available after calling `Session.load` with ``laps=True``
   1051     """
-> 1052     return self._get_property_warn_not_loaded('_laps')

File ~\AppData\Local\Programs\Python\Python310\lib\site-packages\fastf1\core.py:1017, in Session._get_property_warn_not_loaded(self, name)
   1015 d = getattr(self, name, None)
   1016 if d is None:
-> 1017     raise DataNotLoadedError("The data you are trying to access has not "
   1018                              "been loaded yet. See `Session.load`")
   1019 return d

DataNotLoadedError: The data you are trying to access has not been loaded yet. See `Session.load`
```

**Comment by theOehrly** [comment]
Ok, I'll try too look at this soon.

**Reply by NOCTish** [reply]
I'm really grateful. Thank You

**Reply by theOehrly** [reply]
Ok, so it is working for me. Please try to clear the cache and try again. If there were server errors, it might be, that erroneous data was cached and that this data actually persists longer in cache than a correct API response. This might be the reason, why the problem is not going away by itself.
The easiest way to clear the cache is to simply delete the content of the cache folder.

**Reply by NOCTish** [reply]
Hello again!
I tried deleting cache, even using a vpn while running the code
The issue is persistent.
Is there any place where I can download these Core Warning Datas?

**Reply by theOehrly** [reply]
Please run the following, slightly extended version of the example code that you posted above. This will enable debug logging.
Please send me the full output, then I can see what's going wrong. Preferably, attach it as a text file instead of just pasting it here.

```
import logging

logging.basicConfig(level=logging.DEBUG, style='{', format="{module: <8} {levelname: >10} \t{message}")

import fastf1 as ff1


# ff1.Cache.enable_cache('tmp2_cache')

quali_session = ff1.get_session(2022,'Bahrain','Q')
quali_session.load()
quali_session.laps

```

**Reply by NOCTish** [reply]
Greeting! Done it and here is the result
[Result.txt](https://github.com/theOehrly/Fast-F1/files/10952219/Result.txt)

**Reply by theOehrly** [reply]
Ok, so that brings us a small step further. The server actually refuses to send you the data. One thing stands out though, there's no warning that caching isn't enabled but this warning should be there, because caching was never enabled. Did you just not copy that part of the output? It should basically be the very first thing that is printed?
Also, how did you install FastF1?

**Reply by NOCTish** [reply]
First Q: It sent me the error first time I ran the code, but since it was one the first problems I encountered, I just created tmp2 cache so it won't pop the error again
Second Q: I used pip install fastf1 in VSCode Terminal directly and then, I started to install other requirements via Terminal

**Reply by NOCTish** [reply]
- I don't know if this can be the case or not - But I live in Iran hence, there are website which are either restricted by our government or they have restricted iranian IPs from accessing. I don't know if the website which should send me datas is in this category or not.

**Reply by theOehrly** [reply]
Ahh I fear that the issue might be that you live in Iran. Try to open this URL:
[https://livetiming.formula1.com/static/2022/2022-03-20_Bahrain_Grand_Prix/2022-03-19_Qualifying/DriverList.jsonStream](https://livetiming.formula1.com/static/2022/2022-03-20_Bahrain_Grand_Prix/2022-03-19_Qualifying/DriverList.jsonStream)
Your browser should offer you to download the file but I fear that you will simply get a 403 Error message.

**Reply by NOCTish** [reply]
Yup. That was the case
Though (I'm on the phone RN) I can open it via VPN.
Should the data be reachable if I use vpn while running the code?

**Reply by theOehrly** [reply]
If your system's default internet connection goes through the VPN and VPN makes it accessible, then it should work in theory.

**Reply by NOCTish** [reply]
Ok. I'll try and get back at you.

**Reply by NOCTish** [reply]
@theOehrly I have successfully used VPN and I got the quali Laps Session Datas!
Tnx for the help!
as for the final question, is there anywhere where I can get commands for the codes? and datas?

**Reply by theOehrly** [reply]
Great that we managed to solve this. Also, as a hint, after data for a session has been loaded once, you should be able to use FastF1 without the VPN. It should then just use the cached data.

Regarding your question, I don't understand what you mean. Can you clarify?

**Reply by NOCTish** [reply]
Tnx for the hint!
As for the question
With the "quali_session.load()"
We load the session
and then, with wuch commands, we ask the data
"quali_session.laps"
I want to know if there is like a guide book in which tells what code lines are possible to write and get datas
what are the codes for quali, race, FPs, telemetry and etc

**Reply by theOehrly** [reply]
The documentation has some basic tutorials and examples. And also it links to some external tutorials as well. This all should help you to get started.
